// eslint-plugin-userscripts/dist/rules/better-use-match.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var better_use_match_exports = {};
__export(better_use_match_exports, {
  default: () => better_use_match_default
});
module.exports = __toCommonJS(better_use_match_exports);
var import_createValidator = require("../utils/createValidator");
var better_use_match_default = (0, import_createValidator.createValidator)({
  name: "include",
  description: "use @match instead of @include",
  required: false,
  validator: ({ attrVal, context }) => {
    context.report({
      loc: {
        start: {
          line: attrVal.loc.start.line,
          column: 0
        },
        end: attrVal.loc.end
      },
      messageId: "betterUseMatch"
    });
  },
  messages: {
    betterUseMatch: "Using @include is potentially unsafe and may be obsolete in Manifest v3. Please switch to @match."
  }
});
